src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -0,0 +1,240 @@
+            string workingDirectory,
+            string includeRule,
+            IEnumerable<string> excludeRules)
+        {
+            Func<string,bool> includeFilter = CreateMinimatchFilter(context, includeRule, invert: false);
+            Func<string,bool>[] excludeFilters = excludeRules.Select(excludeRule => 
+                CreateMinimatchFilter(context, excludeRule, invert: true)).ToArray();
+            Func<string,bool> filter = (path) => includeFilter(path) && excludeFilters.All(f => f(path));
+            return filter;
+        }
+
+
+        // Given a globby path, figure out where to start enumerating.
+        // Room for optimization here e.g. 
+        // includeGlobPath = /dir/*foo* 
+        // should map to 
+        // enumerateRootPath = /dir/
+        // enumeratePattern = *foo*
+        // enumerateDepth = SearchOption.TopDirectoryOnly
+        //
+        // It's ok to come up with a file-enumeration that includes too much as the glob filter
+        // will filter out the extra, but it's not ok to include too little in the enumeration.
+        internal static void DetermineFileEnumerationFromGlob(
+            string includeGlobPathAbsolute,
+            out string enumerateRootPath,
+            out string enumeratePattern,
+            out SearchOption enumerateDepth)
+        {
+            int firstGlob = includeGlobPathAbsolute.IndexOfAny(GlobChars);
+            bool hasRecursive = includeGlobPathAbsolute.Contains("**", StringComparison.Ordinal);
+
+            // no globbing
+            if (firstGlob < 0)
+            {
+                enumerateRootPath = Path.GetDirectoryName(includeGlobPathAbsolute);
+                enumeratePattern = Path.GetFileName(includeGlobPathAbsolute);
+                enumerateDepth = SearchOption.TopDirectoryOnly;
+            }
+            else
+            {
+                int rootDirLength = includeGlobPathAbsolute.Substring(0,firstGlob).LastIndexOfAny( new [] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar});
+                enumerateRootPath = includeGlobPathAbsolute.Substring(0,rootDirLength);
+                enumeratePattern = "*";
+                enumerateDepth = hasRecursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;
+            }
+        }
+
+        public static Fingerprint EvaluateKeyToFingerprint(
+            AgentTaskPluginExecutionContext context,
+            string filePathRoot,
+            IEnumerable<string> keySegments)
+        {
+            var sha256 = new SHA256Managed();
+
+            string workingDirectoryValue = context.Variables.GetValueOrDefault(
+                "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
+                )?.Value;
+
+            var resolvedSegments = new List<string>();
+
+            foreach (string keySegment in keySegments)
+            {
+                if (keySegment.Equals("*", StringComparison.Ordinal))
```

Comment:
nit: maybe move all these segment validation checks into a `CheckKeySegment` function so it doesn't add to what is already a fairly complex function.